### DISCUSSION TITLE
Waiting for elements that contain animation/movement to not move anymore

### DISCUSSION THREAD

**Discussion by mduivcap** [cypress-io/cypress#27899] [Questions and Help]
On our website we have anchor navigation and once scrolled down a bit on the page, a back-to-top buttons appears which stick to the bottom of the page. 

```HTML
<div class="back-to-top__button-container visible" style="bottom: 24px;">
  <button class="button...
```

UNLESS you're scrolled down all the way to the bottom and the footer is visible, then it should stay above the footer.

```HTML
<div class="back-to-top__button-container visible" style="bottom: 295.625px;">
  <button class="button...
```

So far so good, I've manually tested this functionality and it works like a charm. 
I even found a way to determine if the element is located above the footer.

```Javascript
cy.get("footer").scrollIntoView();
cy.get(".back-to-top__button-container").then(($el) => {
            const buttonBottomPosition = $el[0].getBoundingClientRect().bottom;
            cy.get("footer").then(($footer) => {
                const footerTopPosition = $footer[0].getBoundingClientRect().top;
                expect(buttonBottomPosition).to.be.lt(footerTopPosition);
            });
        });
```

Now comes the tricky part. 
I can't find a way to wait for the element animation. 
The moment we scrolled to the footer this check is performed and the button is still within the footer and not above it so the test fails.
If we add a cy.wait(600); The button moved above the footer, but is still busy with some animation and it might be that it's a bit within the text area itself.
If we add a cy.wait(1000); all animation stopped and if we make a screenshot now we do see it's 100% in position where it should be.
But what is the proper way to wait for this element to move? 
I thought Cypress would wait for element animation, but I can't seem to find to much information about it.
I did find this page, but I have no clue how to incorporate this into the test
https://docs.cypress.io/guides/core-concepts/interacting-with-elements#Animations

**Comment by portyu9** [comment]
Hi @mduivcap 

You don't need to wait a fixed amount of time for the animation.

The main issue is that your position assertion is inside `.then()`. A `.then()` callback runs only once, so you are measuring the button immediately after `scrollIntoView()` while it is still moving.

Use a retryable `.should()` assertion instead:

```js
cy.get('footer').scrollIntoView()

cy.get('.back-to-top__button-container').should(($button) => {
  const buttonBottom =
    $button[0].getBoundingClientRect().bottom

  const footer =
    Cypress.$('footer')[0]

  const footerTop =
    footer.getBoundingClientRect().top

  expect(buttonBottom).to.be.at.most(footerTop)
})
```

Cypress will rerun the `.should()` callback until the button has moved above the footer or the command times out.

This is preferable to:

```js
cy.wait(1000)
```

because the test waits only as long as the application actually needs.

The animation behavior described in the Cypress documentation is slightly different. Cypress checks whether an element is animating as part of **actionability** before interacting with it, for example before `.click()`. It does not mean Cypress automatically waits for every CSS/JavaScript animation before arbitrary DOM measurements or assertions.

So in this case I would wait for the **observable end state** you actually care about—the button being above the footer—rather than waiting for a specific animation duration.

If the application also exposes a deterministic state such as a class indicating that repositioning has completed, asserting that state would also be a good option.
